### PR TITLE
Fix data uploads to Jatos

### PR DESCRIPTION
**Pull request by dschreij**
These changes make sure that collected data is not only upload incrementally (which Jatos currently doesn't handle well) but also all at the end (which Jatos should handle well) with submitResultData(). I also added sourcemaps, because a lot of people on the forum were worried this was an error that they caused. 

Please not that the format of the data sent to Jatos with `submitResultData` has a different format than than the data sent incrementally. The final data uses `{data: [/* individual data items */], context: { /* context data */}}`, so the jatos data import functions of OpenSesame need to be updated too.

### CODE REVIEW

**Comment by dschreij** [comment]
And apologies for the many changes, I am very fond of the autopep8 tool that automatically standardizes formatting, but an effect is often that all lines of a file change. Leave away commit f445364 if you don't want this.

**Comment by smathot** [comment]
There are still two issues.

The first is when the experiment is aborted by the user. In that case, the partial data that was logged during the experiment is followed by a result `dict` with a `data` key. My guess is that this is because of the call to `jatos.endStudy()` when `abortedByUser` is `True`.

The second is when the experiment finishes successfully on a retry. In that case, the results `dict` is strange, in the sense that there is nested one layer to deep. (See attachment.)

[finished-retry.txt](https://github.com/smathot/opensesame-extension-osweb/files/4464681/finished-retry.txt)

[aborted-data.txt](https://github.com/smathot/opensesame-extension-osweb/files/4464677/aborted-data.txt)

**Comment by dschreij** [comment]
I abandoned the previous approach of using alertify to show a notification. Now a modal is shown once an experiment is finished and data transfer is started. This also shows during the first transfer attempt, and thus users have some status feedback if the transfer may take long; data sets are probably very large when they are submitted all at once in the end, so I bet this will be the case more often than not.
If data transfer fails (due to Jatos saying it's too soon), there is another attempt after 3 seconds, with a maximum of 3 retries.

**Comment by kristian-lange** [comment]
Hi guys, before we do all the testing twice: I'm working on this background worker for jatos.js that sends the resultdata. It's (nearly) finished - I just need to do some testing from my side. But I just released a pre-release of JATOS with the new jatos.js. You can still use the old jatos.js for now with jatos-old.js. Please check this and then, maybe your changes here are not necessary anymore. You can update your JATOS via https://your-jatos-domain/jatos?version=v3.5.3.

**Comment by dschreij** [comment]
Thanks @kristian-lange,
There are a lot of other worthwhile improvements in this PR, so I we will merge most of it anyway. I expect the only changes that are needed to work with the new Jatos will take place in our `onFinishedHandler` function, so that is minimal and we'll do that in a new PR once you released the new Jatos version. For now, Sebastiaan and I can keep testing this PR against the old Jatos and see how it behaves.

**Comment by kristian-lange** [comment]
Hi guys, jatos.js (old and new) already has automatic retries if an request failed in many cases, incl the jatos.submitResultData and jatos.appendResultData. It doesn't harm to have an additional retry mechanism in OSWeb' jatos.js but it's strictly not necessary. There are even two fields `jatos.httpRetry` (number of retries) and `jatos.httpRetryWait` (waiting time in ms between each attempt) you can use to configure it a bit.

**Comment by dschreij** [comment]
Hi @kristian-lange, good to know! However this didn't work for us in the case that there was too little time between two consecutive requests. Only an error message is printed to the console then that a new request cannot be sent until the last one is finished. Admittedly, the experiment we used tested an edge case in which an `appendResultData` call was directly followed by the `submitResultData` call, which probably occurs rarely, but still we trying to cover all the holes to make sure that users don't bump into any unexpected behavior, no matter how unlikely.

**Comment by smathot** [comment]
Thanks, guys. I tested it with the following sessions:

- 2 × successful finished sessions of 12 trials. This results in a `FINISHED` state and a complete `dict` with a `data` key.
- 2 × user-aborted sessions (3 trials and 1 trials). This results in a `FAIL` (experimented aborted by user) state and also a complete `dict` with a `data` key.
- 2 × crashed sessions by forcing a bug in the experiment (2 trials an 1 trial). This results in a `FAIL` (Errors occurred. See log for details.) state and a broken (but recoverable) results set.

In other words, this seems to be exactly what we were aiming for!
